Added a check for :z :Z SELinux flags in Volumes

oatmealraisin opened the pull request:
These flags are handled when they are passed to Docker. See [here](http://www.projectatomic.io/blog/2015/06/using-volumes-with-docker-can-cause-problems-with-selinux/)

Fixes #587 

@bparees ptal?

edit:
My branch was pretty far behind. After the rebase, it looks like we don't check for the ':' character in filepaths. This means we need only change the call to strings.LastIndex() to strings.IndexAny(), so that we split on the first ':'

bparees commented on pkg/api/types.go:
```diff
@@ -451,8 +452,16 @@ func (l *VolumeList) Set(value string) error {
+
+	// Trim possible :z/:Z flags before sanity check
+	destination := s.Destination
+	if strings.HasSuffix(destination, ":z") || strings.HasSuffix(destination, ":Z") {
```
are those the only possible flags?

oatmealraisin replied:
After the rebase it doesn't look like we need to check for specific flags, so Docker should return an error if we use a mode it doesn't support

bparees commented on pkg/api/types.go:
```diff
@@ -505,14 +505,15 @@ func (l *VolumeList) Set(value string) error {
 	mount[1] = strings.Trim(mount[1], `"'`)
+
 	s := VolumeSpec{Source: filepath.Clean(mount[0]), Destination: filepath.ToSlash(filepath.Clean(mount[1]))}
 	if IsInvalidFilename(s.Source) || IsInvalidFilename(s.Destination) {
```
so this check doesn't fail?

oatmealraisin replied:
No, since we don't check for ':', both source and destination would pass this check. However, if docker doesn't like either filename, it will fail. For example:
```
 $ s2i build https://github.com/pmorie/simple-ruby openshift/ruby-20-centos7 test-ruby-app --volume="/tmp/asdf:z:/home/bundle"
Your branch is up-to-date with 'origin/master'.
ERROR: An error occurred: Error response from daemon: Invalid bind mount spec "/tmp/asdf:z:/home/bundle": volumeinvalidmode: invalid mode: "/home/bundle"
```
I'll expand the PR description to be more clear

bparees replied:
cool, deferring to docker makes sense to me.

bparees commented:
add a test(positive and negative please) and it should be good to go.

oatmealraisin commented:
@bparees Added tests, ptal?
[test]

bparees commented on pkg/api/types.go:
```diff
@@ -505,14 +505,20 @@ func (l *VolumeList) Set(value string) error {
 	var mount []string
-	pos := strings.LastIndex(value, ":")
+	var pos int
+	if strings.HasPrefix(value, "C:\\") {
```
don't assume it's the C drive....  or that it's uppercased.

i'm not familiar w/ how mounting works when running on windows but you seem to be stripping the prefix before passing the path to docker which seems like the wrong answer.

@csrwng have you dealt w/ this in cluster up at all?  (passing a windows path as a mount path to a container?)

bparees replied:
nm, i see you're not stripping the prefix.  but still the check needs to be looser.

oatmealraisin replied:
I added a regexp match for windows drives.

bparees commented on pkg/api/types.go:
```diff
@@ -504,20 +505,31 @@ func (l *VolumeList) Set(value string) error {
 	}
+
+	var pos int
+	isWindows, _ := regexp.MatchString("^[A-Z]:\\.*", value)
```
what about lower case values?

oatmealraisin replied:
Can Windows drive letters be lowercase?

bparees replied:
yes.  i mean it's basically case insensitive, so users can type "c:\foo" which maps to C:\foo effectively.

oatmealraisin replied:
Oh yeah it is case insensitive, good call

oatmealraisin commented:
Having some issues with the test that expects a failure

oatmealraisin commented:
@bparees Added a check for lowercase Windows drive letters, negated the negative test. ptal?

oatmealraisin commented:
Also, changed all whitespace to tabs in the test case -- there was some inconsistent tabbing :)

bparees commented on hack/test-stirunimage.sh:
```diff
@@ -8,62 +8,62 @@ export PATH="$PWD/_output/local/bin/$(go env GOHOSTOS)/$(go env GOHOSTARCH):$PAT
 function time_now()
 {
-  date +%s000
+	date +%s000
```
did you convert spaces to tabs?  please switch it back.  we try to use spaces everywhere.

oatmealraisin replied:
Changed my vimrc, will do from now on

bparees commented on pkg/api/types_test.go:
```diff
@@ -22,6 +24,7 @@ func TestVolumeListSet(t *testing.T) {
 		{`'/test':"/foo"`, VolumeList{{Source: "/test", Destination: "/foo"}}},
 		{`C:\test:/bar`, VolumeList{{Source: `C:\test`, Destination: "/bar"}}},
 		{`C:\test:bar`, VolumeList{{Source: `C:\test`, Destination: "bar"}}},
+		{`D:\test:bar`, VolumeList{{Source: `D:\test`, Destination: "bar"}}},
```
add a lower case letter test.

oatmealraisin replied:
Also added a C:/ + SELinux test

bparees commented:
> Also, changed all whitespace to tabs in the test case -- there was some inconsistent tabbing :)

unfortunately you went the wrong way :(

we should be converting tabs to spaces when found.

oatmealraisin commented:
@bparees Hopefully I've got everything :) ptal?

bparees commented:
lgtm [merge]

bparees commented:
[merge]


On Wed, Jan 25, 2017 at 4:07 PM, OpenShift Bot <notifications@github.com>
wrote:

> Source To Image Merge Results: FAILURE (https://ci.openshift.redhat.
> com/jenkins/job/test_pr_s2i/415/) (Base Commit: 0fd9583
> <https://github.com/openshift/source-to-image/commit/0fd9583fab8256281d6122cfd846deb0dc360732>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/682#issuecomment-275233026>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEvl3gJxpZc1kdHaU3kTxVqnMzfPOB5Bks5rV7l5gaJpZM4LrovP>
> .
>



-- 
Ben Parees | OpenShift

oatmealraisin commented:
@bparees Pretty odd failure, I don't think I mess w/ that test.
re[test]

bparees commented:
it's a known flake, jim is working on it.


On Wed, Jan 25, 2017 at 4:34 PM, Ryan Murphy <notifications@github.com>
wrote:

> @bparees <https://github.com/bparees> Pretty odd failure, I don't think I
> mess w/ that test.
> re[test]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/682#issuecomment-275239742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEvl3vw0P5PaWv-CouS1sr6q7qoegwBAks5rV7_KgaJpZM4LrovP>
> .
>



-- 
Ben Parees | OpenShift

oatmealraisin commented:
Just found it haha
Flake #533, [test]

bparees commented:
[merge]

bparees commented:
@oatmealraisin

sorry, this is not the flake i thought it was, your test is failing on
windows:

ERROR: An error occurred: Error response from daemon: {"message":"Invalid
bind mount spec \"C:\\\\cygwin\\\\tmp\\\\sti\\\\test-work.xZTj:/home/:z\":
Invalid volume specification:
'C:\\cygwin\\tmp\\sti\\test-work.xZTj:/home/:z'"}

you're ​going to need to make the test a little smarter so it expects
failure on windows (or doesn't run on windows)​


On Wed, Jan 25, 2017 at 5:23 PM, OpenShift Bot <notifications@github.com>
wrote:

> Source To Image Merge Results: FAILURE (https://ci.openshift.redhat.
> com/jenkins/job/test_pr_s2i/418/) (Base Commit: 0fd9583
> <https://github.com/openshift/source-to-image/commit/0fd9583fab8256281d6122cfd846deb0dc360732>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/682#issuecomment-275252032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEvl3pT5ctk7IdD71j2F1x6YBG3XtaY5ks5rV8tKgaJpZM4LrovP>
> .
>



-- 
Ben Parees | OpenShift

oatmealraisin commented:
@bparees That directory looks like it should work? Regardless, I've added a tid-bit to avoid windows.

bparees commented on hack/test-stirunimage.sh:
```diff
@@ -127,12 +127,21 @@ test_debug "s2i build with git proto"
+    check_result $? "${WORK_DIR}/s2i-volume-correct.log"
+    ! s2i build https://github.com/openshift/cakephp-ex openshift/php-55-centos7 test --volume "${S2I_WORK_DIR}:/home/:z:z" &> "${WORK_DIR}/s2i-volume-incorrect.log"
+    check_result $? "${WORK_DIR}/s2i-volume-incorrect.log"
+fi
```
seems like we should have a test that does run on windows too, otherwise --volume isn't getting tested on windows at all.

oatmealraisin commented:
@bparees I think that the test should pass regardless of Windows, because we use the generated work-dir. I tried looking for a reason why it's failing, but couldn't find anything. I've added a specific case for WIndows.

oatmealraisin commented:
@bparees Narrowed it down to only my tests to see if I'm timing out the test, not failing my tests. Could this be a problem? Not sure what would have changed to make the unix test fail as well.

bparees commented on hack/test-stirunimage.sh:
```diff
@@ -71,69 +71,84 @@ trap cleanup EXIT SIGINT
+   test_debug "s2i build with unix volume options"
+   s2i build https://github.com/openshift/cakephp-ex openshift/php-55-centos7 test --volume "${S2I_WORK_DIR}:/home/:z" &> "${WORK_DIR}/s2i-volume-correct.log"
+   check_result $? "${WORK_DIR}/s2i-volume-correct.log"
+   ! s2i build https://github.com/openshift/cakephp-ex openshift/php-55-centos7 test --volume "${S2I_WORK_DIR}:/home/:z:z" &> "${WORK_DIR}/s2i-volume-incorrect.log"
```
given that you're getting one "TEST PASSED" after the "s2i build with unix volume options", it would seem like the test is hanging on this second test.  have you tried running this command by hand?

oatmealraisin replied:
It passes locally, which is the strange part. I'm changing the image/directory it runs on in case that changes things

rhcarvalho commented on pkg/api/types.go:
```diff
@@ -504,20 +505,31 @@ func (l *VolumeList) Set(value string) error {
 	}
+
+	var pos int
+	isWindows, _ := regexp.MatchString("^[A-Za-z]:\\.*", value)
```
Apparently this method `Set` is called in a loop. It would not be a bad idea to compile the regexp only once at initialization time. Also, in general, ignoring errors like here raises a red flag.

Suggestion:

```go
var someName = regexp.MustCompile(`...`)
```

rhcarvalho commented on pkg/api/types.go:
```diff
@@ -504,20 +505,31 @@ func (l *VolumeList) Set(value string) error {
+	var pos int
+	isWindows, _ := regexp.MatchString("^[A-Za-z]:\\.*", value)
+	if isWindows {
+		pos = strings.IndexAny(value[3:], ":") + 3
```
Very suspicious to use `strings.IndexAny` when [`strings.Index`](https://golang.org/pkg/strings/#Index) would do it: we're looking for a single character in the input.

The "magic number" 3 is better given a name, what does it mean?
Seems unclear for people reading the code why on Windows there is this offset.

Also suspicious that we're changing from `strings.LastIndex` to `strings.IndexAny` / `strings.Index`, as they behave quite differently when there are more occurrences of ":" in the input.

oatmealraisin replied:
I believe I just used IndexAny because that was the first to come to mind. I'll change it to `strings.Index`

The '3' is for the `C:\` that prefixes windows directories. This is so that we avoid that detail. I'll add a comment to clarify.

We are changing it to `strings.Index` to grab the first instance of ':', which will preserve the ':z' flag in the destination directory.

rhcarvalho commented on hack/test-stirunimage.sh:
```diff
@@ -8,14 +8,14 @@ export PATH="$PWD/_output/local/bin/$(go env GOHOSTOS)/$(go env GOHOSTARCH):$PAT
 function time_now()
 {
-  date +%s000
+    date +%s000
```
Why are there so many whitespaces changes in the same commit?

It is considered good practice, and helps understanding the project history if changes in a commit are related to the same fix/feature/etc. This commit will be always explained as "SElinux ...".

oatmealraisin replied:
I'm sorry, I noticed that the whitespace in the file was a bit mis-matchy and decided to pipe it through a formatter.

jim-minter replied:
I think it would be preferable to revert the whitespace changes and pursue them in a separate PR.

bparees replied:
i don't really see an issue w/ doing some formatting cleanup when you're already touching a file.  if we force people to create a separate commit/PR it's just going to make people more reluctant to do it (or forget to do it).  Especially since it increases the risk you're going to create a merge conflict with your other PR.

I'm having a hard time envisioning the scenario where this delta baffles someone, or where someone is going to be digging through our git history looking for the commit that reformatted whitespace.

oatmealraisin commented:
@rhcarvalho Thanks for the review! I've updated the commit

oatmealraisin commented:
[test]

rhcarvalho commented on pkg/api/types.go:
```diff
@@ -504,20 +506,32 @@ func (l *VolumeList) Set(value string) error {
+		// check for ':' after that prefix.
+		pos = strings.Index(value[3:], ":") + 3
+	} else {
+		pos = strings.Index(value, ":")
```
Please double check that it is okay to change from `strings.LastIndex` to `strings.Index`.

oatmealraisin replied:
? i'm not sure what I'm supposed to be checking. I'm changing the code to grab the first ':' so that it preserves any possible file mode flags that could suffix the destination directory. Am I missing something? All of the tests pass (besides the added ones)

bparees replied:
changing it to Index is the crux of the fix here.. we want to split on the first colon, not the last colon, because the user can supply a mount string like "/src:/dest:z" and we want to split that into "/src" and "/dest:z", not "/src:/dest" and ":z"

rhcarvalho replied:
My point was that if it was `LastIndex` it must have been for a reason.
IIRC we are now accepting file names that contain `:` (since we added support for Windows).

In Linux and macOS file systems is it totally fine to have `:` in file names:

```
$ touch :hello
$ ls
:hello
```

So in this case, we could have input like `/src-:hello:/dest:z` -- I don't see a test case for that in `pkg/api/types_test.go`.

oatmealraisin replied:
I don't think we can easily handle this. If we can have an arbitrary amount of ':' in src/dest, how could we know where to split the string?

rhcarvalho commented on pkg/api/types.go:
```diff
@@ -504,20 +506,32 @@ func (l *VolumeList) Set(value string) error {
+	if isWindowsRegExp.MatchString(value) {
+		// We need to skip the 'C:\' that prefixes Windows directories, so we only
+		// check for ':' after that prefix.
+		pos = strings.Index(value[3:], ":") + 3
```
The documentation of this function says:
```
This function parses the string that contains source:destination pair.
```

Maybe that needs to be revisited. How would it look like on Windows?

```
C:\Users\John\projects\webapp:/destination/in/container
```

?

oatmealraisin replied:
I've updated the comment to explain the extra functionality

oatmealraisin commented:
@bparees @rhcarvalho Docker doesn't handle invalid/extra file mode flags anymore, for some reason. This has been causing the tests to hang when it gets to the incorrect part. The `Set()` method I've been editing now returns an error when the user passes in something like `/path/to/dir:/other/path/to/dir:rw:z`

Unsure of whether this will fix the windows tests, will see

rhcarvalho commented:
> The Set() method I've been editing now returns an error when the user passes in something like /path/to/dir:/other/path/to/dir:rw:z

The correct syntax is to combine the flags with comma: `/path/to/dir:/other/path/to/dir:rw,z`.

https://docs.docker.com/engine/reference/run/#volume-shared-filesystems

> The **comma-delimited** `options` are [rw|ro], [z|Z], [[r]shared|[r]slave|[r]private], and [nocopy].

jim-minter commented on hack/test-stirunimage.sh:
```diff
@@ -86,6 +86,20 @@ test_debug "s2i build with relative path with file://"
 
+if [[ "$OSTYPE" == "cygwin" ]]; then
+   test_debug "s2i build with windows volume options"
+   s2i build cakephp-ex openshift/php-55-centos7 test --volume "${WORK_DIR}:/home/:z"
```
Missing &> "${WORK_DIR}/s2i-volume-correct.log"

oatmealraisin replied:
Thanks for the review! These are missing to better debug what was happening on the Jenkins job.

jim-minter commented on hack/test-stirunimage.sh:
```diff
@@ -86,6 +86,20 @@ test_debug "s2i build with relative path with file://"
+   test_debug "s2i build with windows volume options"
+   s2i build cakephp-ex openshift/php-55-centos7 test --volume "${WORK_DIR}:/home/:z"
+   check_result $? "${WORK_DIR}/s2i-volume-correct.log"
+   ! s2i build cakephp-ex openshift/php-55-centos7 test --volume "${WORK_DIR}:/home/"
```
Missing &> "${WORK_DIR}/s2i-volume-incorrect.log"

jim-minter replied:
I don't love the leading `!` - perhaps check_result could be extended to deal with expected negatives?

jim-minter replied:
Why is this test expected to fail?  Is not providing a `:z` flag not allowed?

oatmealraisin replied:
Like above, some tests were changed to debug the windows job

jim-minter commented on hack/test-stirunimage.sh:
```diff
@@ -86,6 +86,20 @@ test_debug "s2i build with relative path with file://"
 s2i build file://./cakephp-ex openshift/php-55-centos7 test --loglevel=5 &> "${WORK_DIR}/s2i-rel-proto.log"
 check_result $? "${WORK_DIR}/s2i-rel-proto.log"
 
+if [[ "$OSTYPE" == "cygwin" ]]; then
```
I think that if you use $S2I_WORK_DIR instead of $WORK_DIR, you should not need this if/then/else.
On non-Windows systems, S2I_WORK_DIR equals WORK_DIR.  On Windows systems, S2I_WORK_DIR translates foo/bar to C:\path\to\foo\bar for you.

jim-minter commented:
Unfortunately, I don't think this PR quite makes sense yet as it stands as far as s2i on Windows is concerned.

If you're running s2i on windows, I suppose it's because you're talking to a docker daemon running on Linux.  Any bind using -v is inevitably of a folder locally on the Linux box, into the container running on the Linux box.  Therefore trying to bind anything of the form c:\foo\bar into a Linux container is meaningless.

It's somewhat fiddly - running s2i on Windows:
- the path to a repo (if not a URL) will be of the form C:\foo\bar
- the argument for an injection will be of the form C:\foo\bar:/baz/bash
- the argument for a bound volume will be /foo/bar:/baz/bash[:flags]

jim-minter commented on hack/test-stirunimage.sh:
```diff
@@ -86,6 +86,20 @@ test_debug "s2i build with relative path with file://"
+   test_debug "s2i build with unix volume options"
+   s2i build cakephp-ex openshift/php-55-centos7 test --volume "${S2I_WORK_DIR}:/home/:z" --loglevel=5 &> "${WORK_DIR}/s2i-volume-correct.log"
+   check_result $? "${WORK_DIR}/s2i-volume-correct.log"
+   ! s2i build cakephp-ex openshift/php-55-centos7 test --volume "${S2I_WORK_DIR}:/home/:z:z" --loglevel=5 &> "${WORK_DIR}/s2i-volume-incorrect.log"
```
Couldn't this be done by unit test?

csrwng commented:
@jim-minter Docker for Windows does let you mount windows directories into your running containers with -v like: `-v c:\Users\cewong\mydir:/var/mydir`

jim-minter commented:
@csrwng urg - how to tell if you're running Docker for Windows, or talking to a Docker socket on a remote Linux box?

jim-minter commented:
Hmm, if that's the case, then does that imply we really shouldn't attempt to parse any -v argument, but just send it straight through to the Docker layer?

The Windows CI test setup we have won't support a test of the form -v c:\foo\bar:/baz/bash because it doesn't run Docker in a VM.  As a result I'd probably make the testing of this non-Windows only.

csrwng commented:
@jim-minter hmm, that's not easy ... technically Docker for Windows *is* using a remote Linux vm running on Hyper-V. I guess one way would be to look at the DOCKER_HOST value. In the case of Docker for Windows, it's a windows named pipe (as opposed to a tcp host).

bparees commented:
> Hmm, if that's the case, then does that imply we really shouldn't attempt to parse any -v argument, but just send it straight through to the Docker layer?

+1

oatmealraisin commented:
> Hmm, if that's the case, then does that imply we really shouldn't attempt to parse any -v argument, but just send it straight through to the Docker layer?

@bparees @jim-minter I'm a bit confused on what you mean. We need to parse -v to create a VolumeSpec. Would we just get rid of that? Or do you mean we should just find the ':' to split on and leave it there?

jim-minter commented:
@oatmealraisin as a starting point I'd at least see what it looks like if you stop using VolumeList (= []VolumeSpec) entirely for -v and use an opaque []string (but keep using VolumeSpecs for injections, because fundamentally we have to be able to parse those).

Don't forget to fix up `BuildVolumes VolumeList` in `type Config` in pkg/api/types.go.

Looking at the origin code which calls us in this area, I think it shouldn't be problematic for it to pass []string rather than a []VolumeSpec.  See copyToVolumeList in pkg/build/builder/sti.go.

oatmealraisin commented:
@jim-minter @bparees  Sorry for the delay, I was making this more complicated than it was. I've reverted VolumeList and changed BuildVolumes to a []string
However, since we let Docker handle more things, it hangs when an extra :rwx,zZ. Because of this, I've removed the negative test in test-stirunimage.sh. Also, I've changed the test to mount ${STI_WORK_DIR}, so that it will test different directories depending on which Jenkins job is running.

bparees commented:
[merge]

php-coder commented on pkg/api/types.go:
```diff
@@ -232,7 +232,7 @@ type Config struct {
 	// BuildVolumes specifies a list of volumes to mount to container running the
 	// build.
-	BuildVolumes VolumeList
+	BuildVolumes []string
```
When we used `VolumeList` we had [validation for file names](https://github.com/openshift/source-to-image/blob/4dd77215907d810f8fbc9c23dd8c7454f89131eb/pkg/api/types.go#L499-L522). With this change we don't have them anymore. Are we sure that it won't cause problem, especially security-related?

bparees replied:
we're deferring to docker.

if you're using s2i directly, there is no security.

oatmealraisin commented:
@bparees Thanks for merging this, I meant to tag you over the weekend.
